### PR TITLE
Add VerticalSizer for track header

### PR DESCRIPTION
**Pull request by pierricgimmig**
Add vertical sizer class to resize new track header. The cursor
changes to horizontal arrows when the mouse is in the sizer.
We keep a minimum width equal to the height of the timeline
for symmetry, to keep the triangle toggles visible, and to keep 
the track moving feature available. 

Also, set right margin to 0, but keep logic just in case, and tweak 
distance between timeline and tracks for more stream lined look.

image

image

### CODE REVIEW

File: src/OrbitGl/VerticalSizer.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  if (QGuiApplication::instance() != nullptr) {
+    QGuiApplication::restoreOverrideCursor();
+  }
+  RequestUpdate(RequestUpdateScope::kDraw);
+  return event_result;
+}
+
+void VerticalSizer::OnDrag(int x, int y) {
+  on_drag_callback_(x, y);
+  RequestUpdate(RequestUpdateScope::kDraw);
```
Comment (reichlfl) [inline src/OrbitGl/VerticalSizer.cpp:48]: I would argue that in this case, `RequestUpdate` should not happen as well. It's the responsibility of the code in `on_drag_callback_` to decide wether an update is needed or not as the sizer is not updating it's own position.

File: src/OrbitGl/TimeGraph.cpp [hunk 2]
```diff
@@ -887,6 +895,13 @@ void TimeGraph::DoDraw(orbit_gl::PrimitiveAssembler& primitive_assembler,
   if (layout_->GetDrawTimeGraphMasks()) {
     DrawMarginsBetweenChildren(primitive_assembler);
   }
+
+  // Track header sizer.
+  float sizer_width = layout_->GetSpaceBetweenTracks();
+  float header_width = layout_->GetTrackHeaderWidth();
+  track_header_sizer_->SetWidth(sizer_width);
+  track_header_sizer_->SetHeight(GetHeight());
+  track_header_sizer_->SetPos(GetPos()[0] + header_width - sizer_width, GetPos()[1]);
```
Comment (reichlfl) [inline src/OrbitGl/TimeGraph.cpp:904]: `DoDraw` and `DoUpdatePrimitives` must not call any methods that result in a redraw. Move this to `TimeGraph::DoUpdateLayout`.

**Review by reichlfl — CHANGES_REQUESTED** [review]
This is really cool, thanks! I left a couple of comments regarding update / redraw conventions that need to be followed, otherwise we risk some changes not showing up until the next frame.

File: src/OrbitGl/TimeGraph.cpp (hunk 2 above)
Comment (pierricgimmig) [reply src/OrbitGl/TimeGraph.cpp]: Done

File: src/OrbitGl/VerticalSizer.cpp (hunk 1 above)
Comment (pierricgimmig) [reply src/OrbitGl/VerticalSizer.cpp]: Done

File: src/OrbitGl/VerticalSizer.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+VerticalSizer::VerticalSizer(CaptureViewElement* parent, const orbit_gl::Viewport* viewport,
+                             const TimeGraphLayout* layout,
+                             std::function<void(int /*x*/, int /*y*/)> on_drag_callback)
+    : CaptureViewElement(parent, viewport, layout),
+      on_drag_callback_(std::move(on_drag_callback)) {}
+
+void VerticalSizer::DoDraw(PrimitiveAssembler& primitive_assembler, TextRenderer& text_renderer,
+                           const DrawContext& draw_context) {
+  CaptureViewElement::DoDraw(primitive_assembler, text_renderer, draw_context);
+  Quad right_margin_box = MakeBox(GetPos(), GetSize());
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/OrbitGl/VerticalSizer.cpp]: copy-paste is not my friend, changed to `box`.

File: src/OrbitGl/include/OrbitGl/VerticalSizer.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+#include "OrbitGl/CaptureViewElement.h"
+#include "OrbitGl/PrimitiveAssembler.h"
+#include "OrbitGl/TextRenderer.h"
+#include "OrbitGl/TimeGraphLayout.h"
+#include "OrbitGl/Viewport.h"
+
+namespace orbit_gl {
+
+class VerticalSizer : public orbit_gl::CaptureViewElement,
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/OrbitGl/include/OrbitGl/VerticalSizer.h:18]: Done.

File: src/OrbitGl/VerticalSizer.cpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  RequestUpdate(RequestUpdateScope::kDraw);
+  return event_result;
+}
+
+CaptureViewElement::EventResult VerticalSizer::OnMouseLeave() {
+  EventResult event_result = CaptureViewElement::OnMouseLeave();
+  if (QGuiApplication::instance() != nullptr) {
+    QGuiApplication::restoreOverrideCursor();
+  }
+  RequestUpdate(RequestUpdateScope::kDraw);
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/OrbitGl/VerticalSizer.cpp]: Removed

File: src/OrbitGl/VerticalSizer.cpp [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  primitive_assembler.AddBox(right_margin_box, GlCanvas::kZValueMargin, GlCanvas::kBackgroundColor,
+                             shared_from_this());
+}
+
+CaptureViewElement::EventResult VerticalSizer::OnMouseEnter() {
+  EventResult event_result = CaptureViewElement::OnMouseEnter();
+  if (QGuiApplication::instance() != nullptr) {
+    QGuiApplication::setOverrideCursor(Qt::SizeHorCursor);
+  }
+  RequestUpdate(RequestUpdateScope::kDraw);
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/OrbitGl/VerticalSizer.cpp]: Removed

File: src/OrbitGl/include/OrbitGl/VerticalSizer.h [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  explicit VerticalSizer(CaptureViewElement* parent, const orbit_gl::Viewport* viewport,
+                         const TimeGraphLayout* layout,
+                         std::function<void(int /*x*/, int /*y*/)> on_drag_callback);
+  ~VerticalSizer() override = default;
+
+  // Pickable
+  void OnDrag(int /*x*/, int /*y*/) override;
+
+  [[nodiscard]] float GetHeight() const override { return height_; }
+  void SetHeight(float height) { height_ = height; }
```
[the remark this answers is not in the document]
Comment (pierricgimmig) [reply src/OrbitGl/include/OrbitGl/VerticalSizer.h]: Done. Btw, there is an asymmetry with the SetWidth method which does this boilerplate, we should probably fix SetHeight so it behaves similarly.

**Comment by pierricgimmig** [comment]
> This is really cool, thanks! I left a couple of comments regarding update / redraw conventions that need to be followed, otherwise we risk some changes not showing up until the next frame.

Thanks for the review! I addressed all the comments, please take another look. Cheers!

File: src/OrbitGl/include/OrbitGl/TimeGraph.h [hunk 8]
```diff
@@ -200,6 +204,7 @@ class TimeGraph : public orbit_gl::CaptureViewElement, public orbit_gl::Timeline
 
   std::shared_ptr<orbit_gl::GlSlider> horizontal_slider_;
   std::shared_ptr<orbit_gl::GlSlider> vertical_slider_;
+  std::shared_ptr<orbit_gl::VerticalSizer> track_header_sizer_;
```
Comment (florian-kuebler) [inline src/OrbitGl/include/OrbitGl/TimeGraph.h:207]: Why does it need to be a shared pointer?

**Review by florian-kuebler — APPROVED** [review]
LGTM. Only one question.

File: src/OrbitGl/include/OrbitGl/TimeGraph.h (hunk 8 above)
Comment (pierricgimmig) [reply src/OrbitGl/include/OrbitGl/TimeGraph.h:207]: I think because the picking system expects it.

Comment (florian-kuebler) [reply src/OrbitGl/include/OrbitGl/TimeGraph.h:207]: Ah I see. There it probably also is added. Missed that. Thanks

Comment (reichlfl) [reply src/OrbitGl/include/OrbitGl/TimeGraph.h:207]: Yeah, this is needed due to picking, we may want to clean that up...

**Review by reichlfl — APPROVED** [review]
Thanks for the changes, LGTM!

**Pull request merged**